feat: upstream provider and forked network tools [APE-1485]

antazoey opened the pull request:
### What I did

realizing it was absolute pain to need to do anything with the upstream provider, even at the plugin level.
This aids that by adding properties and such to a new class for forked networks and a method for getting that context as well.

- [ ] show case in ape-safe and ape-foundry
- [ ] update in all ecosystem plugins

### How I did it

<!-- Discuss the thought process behind the change -->

### How to verify it

<!-- Discuss any methods that should be used to verify the change -->

### Checklist

<!-- All PRs must complete the following checklist before being merged -->

- [ ] All changes are completed
- [ ] New test cases have been added
- [ ] Documentation has been updated

fubuloubu commented on src/ape/api/networks.py:
```diff
@@ -1046,6 +1046,54 @@ def verify_chain_id(self, chain_id: int):
+        """
+
+        config_choice = self._network_config.get("upstream_provider")
+        if provider_name := config_choice or self.upstream_network.default_provider:
```
shouldn't it always have a default?

antazoey replied:
`.default_provider` is optional!

Let's say for example you are forking some network that does not have any providers associated with it. It would fail in that case.

I am not sure that is very easy in today's ecosystem... but with the plugin system, anything is possible.

antazoey replied:
We could raise an error instead and change this to not be Optional? But technically, None fits the bill

fubuloubu replied:
Yeah, I think the corner case that you have a network with no providers then it should just raise

antazoey replied:
Ape does ship with networks without providers, so it is quite normal! But none of the upstream providers..
All the forked networks themselves are provider-less.

antazoey replied:
done

fubuloubu reviewed (approved):
love this

antazoey commented:
I addressed feedback and need to merge because my I need for other work and is hard to manage without!
If more feedback arises, please let me know :)
